File: src/src_platform_ava/fmg/ava/img/Animator.cs
using System;
using System.Collections.Generic;
using System.Reactive.Linq;
using Avalonia.Animation;

namespace fmg.ava.img {

    public class Animator : fmg.core.img.IAnimator, IDisposable {

        private class SubscribeInfo {
            public bool active = true;    // enabled?
            public DateTime startTime = DateTime.Now; // start time of subscribe
            public Action<TimeSpan /* time from the beginning of the subscription */> callback;
        }
        private readonly IDisposable _unsubscribe;
        private readonly IDictionary<object /* subscriber */, SubscribeInfo> _subscribers;
        private readonly Animatable animatable;

        private static Animator _singleton;
        public static Animator Singleton { get { // not synchronized. since should work only in the thread of the UI.
            if (_singleton == null)
                _singleton = new Animator();
            return _singleton;
        } }

        private Animator() {
            _subscribers = new Dictionary<object, SubscribeInfo>();
            animatable = new Animatable {
                Clock = new Clock()
            };
            _unsubscribe = animatable.Clock.Subscribe(
                value => {
                    var currentTime = DateTime.Now;
                    foreach (var item in _subscribers) {
                        var subscribeInfo = item.Value;
                        if (subscribeInfo.active)
                            subscribeInfo.callback(currentTime - subscribeInfo.startTime);
                    }
                },
                ex => { System.Diagnostics.Debug.Assert(false, ex.ToString()); },
                () => { System.Diagnostics.Debug.WriteLine("Closed"); }
            );
        }

        public void Subscribe(object subscriber, Action<TimeSpan /* time from start subscribe */> subscriberCallbackMethod) {
            if (!_subscribers.ContainsKey(subscriber)) {
                var info = new SubscribeInfo {
                    callback = subscriberCallbackMethod
                };
                _subscribers.Add(subscriber, info);
            } else {
                var info = _subscribers[subscriber];
                info.active = true;
                info.startTime = DateTime.MinValue + (DateTime.Now - info.startTime); // apply of pause delta time
            }
        }

        public void Pause(object subscriber) {
            SubscribeInfo info = _subscribers[subscriber];
            if (info == null)
                return;
            info.active = false;
            info.startTime = DateTime.MinValue + (DateTime.Now - info.startTime); // set of pause delta time
        }

        public void Unsubscribe(object subscriber) {
            _subscribers.Remove(subscriber);
        }

        public void Dispose() {
            _unsubscribe.Dispose();
            _subscribers.Clear();
            animatable.Clock = null;
        }

    }

}

File: FastMines_jfx/src/main/java/fmg/jfx/img/Animator.java
package fmg.jfx.img;

import java.util.Date;
import java.util.HashMap;
import java.util.Map;
import java.util.function.Consumer;

import javafx.animation.AnimationTimer;

import fmg.core.img.IAnimator;

public class Animator implements IAnimator, AutoCloseable {

    private static class SubscribeInfo {
        public boolean active = true;    // enabled?
        public long    startTime = new Date().getTime(); // start time of subscribe
        public Consumer<Long /* time from the beginning of the subscription */> callback;
    }
    /** one timer to all subscribers */
    private final AnimationTimer timer;
    private final Map<Object /* subscriber */, SubscribeInfo> subscribers;

    private static Animator singleton;
    public static Animator getSingleton() { // not synchronized. since should work only in the thread of the UI.
        if (singleton == null)
            singleton = new Animator();
        return singleton;
    }

    private Animator() {
        subscribers = new HashMap<>();
        timer = new AnimationTimer() {

            @Override
            public void handle(long now) {
                long currentTime = new Date().getTime();
                subscribers.forEach((k, v) -> {
                    if (v.active)
                        v.callback.accept(currentTime - v.startTime);
                });
            }

        };
        timer.start();
    }

    @Override
    public void subscribe(Object subscriber, Consumer<Long /* time from start subscribe */> subscriberCallback) {
        SubscribeInfo info = subscribers.get(subscriber);
        if (info == null) {
            info = new SubscribeInfo();
            info.callback = subscriberCallback;
            subscribers.put(subscriber, info);
        } else {
            if (info.active)
                return; // alredy run

            info.active = true;
            info.startTime = new Date().getTime() - info.startTime; // apply of pause delta time
        }
    }

    @Override
    public void pause(Object subscriber) {
        SubscribeInfo info = subscribers.get(subscriber);
        if (info == null)
            return;

        if (!info.active)
            return; // alredy paused

        info.active = false;
        info.startTime = new Date().getTime() - info.startTime; // set of pause delta time
    }

    @Override
    public void unsubscribe(Object subscriber) {
        subscribers.remove(subscriber);
    }

    @Override
    public void close() {
        timer.stop();
        subscribers.clear();
    }

}
